miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/WebConfig.java

```diff
@@ -70,4 +91,67 @@ public RequestMappingHandlerAdapter requestMappingHandlerAdapter() {
   public Map<ProgressStep.FactoryType, ProgressStepFactory> progressStepFactoryMap() {
     return progressStepFactories.stream().collect(Collectors.toMap(ProgressStepFactory::getFactoryType, Function.identity()));
   }
+
+  public class ExternalUriBuilder {
+    private final Map<String, Map<String, String>> runUris = new TreeMap<>();
+
+    public Map<String, String> getUris(Run run) {
+      if (run.getId() == Run.UNSAVED_ID || runUris.isEmpty()) return Collections.emptyMap();
+
+      return runUris.get(run.getPlatformType().getKey())
+          .entrySet().stream()
+          .collect(Collectors.toMap(Map.Entry::getKey, m -> expandRunUrl(m.getValue(), run)));
+    }
+
+    private String expandRunUrl(String uriWithPlaceholders, Run run) {
+      return uriWithPlaceholders.replaceAll("\\{id\\}", String.valueOf(run.getId()))
+          .replaceAll("\\{name\\}", run.getName())
+          .replaceAll("\\{alias\\}", run.getAlias());
+    }
+
+    public void setRunReportLinksConfig(String runReportLinksConfigLine) {
+      processLinksConfig(runReportLinksConfigLine, runUris);
+    }
+
+    public void processLinksConfig(String linksConfigLine, Map<String, Map<String, String>> uriMap) {
+      if (LimsUtils.isStringBlankOrNull(linksConfigLine)) return;
+      
+      // linksConfigLine format: <PlatformType,PlatformType>|<link text>|<URI with placeholders>
+      // placeholders can be any of {id}, {name}, {alias}.
+      // multiple run report links can be tilde-separated (~).
+      String[] configStrings = linksConfigLine.split("~");
+      for (int i = 0; i < configStrings.length; i++) {
+        String[] configParts = configStrings[i].split("\\|");
+        if (configParts.length != 3) {
+          throw new IllegalArgumentException(
+              String.format("Invalid configuration: expected three link config parts separated by '|' for config string %s but got %d",
+                  configStrings[i], configParts.length));
+        }
+        
+        Set<String> platformTypeStrings = Sets.newHashSet(configParts[0].trim().split(","));
+        Set<PlatformType> platformTypes = platformTypeStrings.stream().map(pt -> PlatformType.get(pt.trim())).collect(Collectors.toSet());
+        if (platformTypes.isEmpty()) {
+          throw new IllegalArgumentException(String.format("Invalid configuration: could not find any matching platforms for string %s", configParts[0]));
+        }
+        String linkText = configParts[1].trim();
+        String uriWithPlaceholders = configParts[2].trim();
+        try {
+          String validateableUri = uriWithPlaceholders.replaceAll("\\{name\\}", "REPLACE")
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "REPLACE" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)